File: sdk/python/src/securenotify/managers/key_manager.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 KirkyX. All rights reserved.

from typing import Optional, List, Dict, Any
from datetime import datetime

from securenotify.types.api import (
    RegisterPublicKeyRequest,
    RegisterPublicKeyResponse,
    PublicKeyInfo,
)
from securenotify.utils.helpers import parse_datetime
from .base import BaseManager


class KeyManager(BaseManager):
    async def register(
        self,
        public_key: str,
        algorithm: str,
        expires_in: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> RegisterPublicKeyResponse:
        """Register a new public key.

        Raises:
            ValueError: If public_key or algorithm is empty.
            SecureNotifyApiError: On API error.
        """
        request = RegisterPublicKeyRequest(
            public_key=public_key,
            algorithm=algorithm,
            expires_in=expires_in,
            metadata=metadata,
        )
        return await self._execute("register_public_key", request)

    async def get(self, key_id: str) -> PublicKeyInfo:
        data = await self._execute("get_public_key", key_id)
        return PublicKeyInfo(
            id=data["id"],
            channel_id=data["channel_id"],
            public_key=data["public_key"],
            algorithm=data["algorithm"],
            created_at=parse_datetime(data, "created_at"),
            expires_at=parse_datetime(data, "expires_at"),
            last_used_at=parse_datetime(data, "last_used_at"),
            metadata=data.get("metadata"),
        )

    async def list(self) -> List[PublicKeyInfo]:
        data = await self._execute("list_public_keys")
        keys = []
        for item in data.get("keys", []):
            keys.append(
                PublicKeyInfo(
                    id=item["id"],
                    channel_id=item["channel_id"],
                    public_key=item["public_key"],
                    algorithm=item["algorithm"],
                    created_at=parse_datetime(item, "created_at"),
                    expires_at=parse_datetime(item, "expires_at"),
                    last_used_at=parse_datetime(item, "last_used_at"),
                    metadata=item.get("metadata"),
                )
            )
        return keys

    async def revoke(self, key_id: str, reason: Optional[str] = None) -> Dict[str, Any]:
        return await self._execute("revoke_public_key", key_id, reason)

File: sdk/java/src/main/java/securenotify/managers/KeyManager.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

package securenotify.managers;

import securenotify.types.ApiRequest;
import securenotify.types.ApiResponse;
import securenotify.types.PublicKeyInfo;
import securenotify.utils.HttpClient;
import securenotify.utils.RetryHandler;

import java.util.Map;

/**
 * Manager for public key operations.
 */
public class KeyManager {

    private final HttpClient httpClient;
    private final RetryHandler retryHandler;

    public KeyManager(HttpClient httpClient) {
        this(httpClient, RetryHandler.DEFAULT);
    }

    public KeyManager(HttpClient httpClient, RetryHandler retryHandler) {
        this.httpClient = httpClient;
        this.retryHandler = retryHandler;
    }

    /**
     * @param publicKey The PEM-encoded public key
     * @param algorithm The encryption algorithm (RSA-2048, RSA-4096, ECC-SECP256K1)
     * @param expiresIn Optional expiration time in seconds
     */
    public PublicKeyInfo register(String publicKey, String algorithm, Integer expiresIn, Map<String, Object> metadata) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .publicKey(publicKey)
                .algorithm(algorithm)
                .expiresIn(expiresIn)
                .metadata(metadata)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.post("api/register", request, PublicKeyInfo.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public PublicKeyInfo register(String publicKey) throws Exception {
        return register(publicKey, "RSA-4096", null, null);
    }

    public PublicKeyInfo get(String keyId) throws Exception {
        return retryHandler.execute(() -> {
            try {
                return httpClient.get("api/keys/" + keyId, null, PublicKeyInfo.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public PublicKeyInfo getByChannel(String channelId) throws Exception {
        return retryHandler.execute(() -> {
            try {
                return httpClient.get("api/register", Map.of("channelId", channelId), PublicKeyInfo.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    /**
     * @param confirmationHours Hours to wait for confirmation
     */
    public ApiResponse<?> revoke(String keyId, String reason, Integer confirmationHours) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .reason(reason)
                .confirmationHours(confirmationHours)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.delete("api/keys/" + keyId, request, Map.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    public ApiResponse<?> revoke(String keyId, String reason) throws Exception {
        return revoke(keyId, reason, 24);
    }
}
